torch/_ops.py

```diff
@@ -788,11 +786,77 @@ def tags(self):
 # TorchBindOpOverload are those custom ops which have at least one overload's
 # schema consists of torch.ScriptObject (i.e. custom class) input.
 # TorchBindOpOverload will skip C++ dispatcher and purely dispatched in python
-# when its inputs contain FakeScriptObject.
+# when its inputs contain FakeScriptObject in a similar way as higher order ops.
 class TorchBindOpOverload(OpOverload):
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
-        self._is_torchbind_op = True
+
+        self.non_fallthrough_keys = torch._C._dispatch_keyset_full()
+
+        _TORCH_BIND_OP_DEFAULT_FALLTHROUGH_DISPATCH_KEYS = [
+            DispatchKey.PythonDispatcher,  # type: ignore[attr-defined]
+            DispatchKey.PythonTLSSnapshot,  # type: ignore[attr-defined]
+            DispatchKey.ADInplaceOrView,
+            DispatchKey.BackendSelect,
+        ]
+        for key in _TORCH_BIND_OP_DEFAULT_FALLTHROUGH_DISPATCH_KEYS:
+            self.non_fallthrough_keys = self.non_fallthrough_keys.remove(key)
+
+    # Note: we automaticallly add implementations for modes that useful for tracing.
+    # We only add implementation when we must dispatch in python by checking the inputs.
+    # We cannot do it before seeing the inputs ecause some torch bind ops
+    # (e.g. profiler record function) might be registered globally
+    # with a single default implementation in cpp. We might accidently trace them
+    # into graph.
+    def _maybe_py_impl_default_dispatch_mode(self):
+        def wrapper(mode, *args, **kwargs):
+            return _mannually_invoke_dispatch_mode_in_python(
+                mode, self, *args, **kwargs
+            )
+
+        for mode in (
+            torch._subclasses.functional_tensor.FunctionalTensorMode,
+            torch.fx.experimental.proxy_tensor.ProxyTorchDispatchMode,
+            torch._subclasses.fake_tensor.FakeTensorMode,
+        ):
+            if mode not in self.python_key_mode_table:
+                self.py_impl(mode)(wrapper)
+
+    # use `self_` to avoid naming collide with arguments that
+    # are named "self". This way, they can be called by kwargs.
+    def __call__(self_, *args, **kwargs):  # noqa: B902
+        # The path when any inputs are FakeScriptObject, we need to
+        # skip c++ dispatcher and dispatch in python through _get_dispatch of python_dispatcher.
+        if _must_dispatch_in_python(args, kwargs):
+            self_._maybe_py_impl_default_dispatch_mode()
+            return self_.dispatch_in_python(args, kwargs)
+
+        return self_._op(*args, **kwargs)
+
+    def dispatch_in_python(self, args, kwargs):
+        dispatch_key_set = _compute_keyset(args, kwargs, self.non_fallthrough_keys)
```

Comment:
This is a little tricky because there's already a source of truth for non_fallthrough_keys in C++. Compare this with the HOP case, where the entire source of truth is in Python because HOPs don't go through C++.

self.non_fallthrough_keys should come from:
- nonFallthroughKeys from C++ (the OperatorEntry has its own nonFallthroughKeys)
- This key set should then be updated for any .py_impl(non_fallthrough), .py_impl(fallthrough) that we add for the operator.

I wouldn't worry about caching this for now (also because open registration means that if we were to cache we'd need to think about cache invalidation)